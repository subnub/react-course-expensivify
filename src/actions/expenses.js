import uuid from "uuid"
import {database} from "../firebase/firebase"


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type:"ADD_EXPENSE",
    expense: expense
})

export const startAddExpense = (expenseData = {}) => {

    return (dispatch, getState) => {

        const {desc = "", note="", amount=0, createdAt=0} = expenseData;
        const expense = {desc, note, amount, createdAt}
        return database.ref().child("users").child(getState().auth.uid).child("expenses").push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }

    

    

}

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: "REMOVE_EXPENSE",
    id
})

export const startRemoveExpense = ({id} = {}) => {

    return (dispatch, getState) => {
        return database.ref().child("users").child(getState().auth.uid).child("expenses").child(id).remove().then(() => {
            dispatch(removeExpense({id}))
        })
    }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
})

export const startEditExpense = (id, updates) => {

    return (dispatch, getState) => {
        database.ref().child("users").child(getState().auth.uid).child("expenses").child(id).set({...updates}).then(() => {
            dispatch(editExpense(id, updates));
        })
    }
}

export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
})

export const startSetExpenses = () => {

    return (dispatch, getState) => {
        return database.ref().child("users").child(getState().auth.uid).child("expenses").once("value").then((snapshot) => {

            const dataExpenses = []

            snapshot.forEach((data) => {
                
                dataExpenses.push({
                    id: data.key,
                    ...data.val()
                });
            })
            
            dispatch(setExpenses(dataExpenses))
        })
    }
}
