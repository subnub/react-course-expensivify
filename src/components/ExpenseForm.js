import React from "react"
import moment from "moment";
import {SingleDatePicker} from "react-dates";

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            desc: props.expense ? props.expense.desc : "",
            noteText: props.expense ? props.expense.note : "",
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calanderFocused: false,
            errorState: ""
        }
    }

    

    onDescChange = (e) => {
        const desc = e.target.value;
        this.setState(() => ({desc}))
    }

    onNoteChange = (e) => {
        const noteText = e.target.value;
        this.setState(() => ({noteText}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            
            this.setState(() => ({amount}))
        }
    }

    onDateChange = (createdAt) => {

        if (createdAt) {
            this.setState(() => ({createdAt}))
        }
    }

    onFocusChange = ({focused}) => {

        this.setState(() => ({calanderFocused: focused}))
    }

    onSubmit = (e) => {

        e.preventDefault();

        if (!this.state.desc || !this.state.amount) {
            // Set Error state 
            this.setState(() => ({errorState: "Please Provide Description and amount"}))
        } else {
            // Clear Error
            this.setState(() => ({errorState: ""}))
            this.props.onSubmit({
                desc: this.state.desc,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.noteText
            })

        }
    }

    render() {
        return (
            <div>
                {this.state.errorState !== "" && <p>{this.state.errorState}</p>}
                <form onSubmit={this.onSubmit}>
                <input type="text" 
                    placeholder="Description" 
                    autoFocus value={this.state.desc}
                    onChange={this.onDescChange}
                />
                <input 
                    type="text" 
                    placeholder="amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calanderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                />
                <textarea 
                    value={this.state.noteText} 
                    placeholder="Add a note to your expense (optional)"
                    onChange={this.onNoteChange}>
                </textarea>
                <button>Add Expense</button>
                </form>
            </div>
        )
    }
}