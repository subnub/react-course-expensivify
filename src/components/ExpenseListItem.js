import React from "react";
import {Link} from "react-router-dom"
import moment from "moment"
import numeral from "numeral"

const ExpenseListItem = ({desc, amount, createdAt, id}) => (

        <Link className="list-item" to={`/edit/${id}`}>
        <div>
        <h3 className="list-item__title">{desc}</h3>
        <span className="list-item__subtotal">{numeral(amount / 100).format("$0,0.00")} </span>
        </div>
        
        <h3 className="list-item__data">{moment(createdAt).format("MMMM Do YYYY")}</h3>
        
        </Link>   
  
)

export default ExpenseListItem;