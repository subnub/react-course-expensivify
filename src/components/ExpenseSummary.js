import React from "react";
import {connect} from "react-redux"
import selectExpenses from "../selectors/expense"
import expenseTotal from "../selectors/expenseTotal"
import numeral from "numeral"
import {Link} from "react-router-dom";


export const ExpensesSummary = (props) => {

    const total = expenseTotal(props.expenses)
    const totalFormated = numeral(total / 100).format("$0,0.00")
    const expenseLength = props.expenses.length;
    const viewingText = expenseLength === 1 ? "1 expense" : expenseLength + " expenses"

    return (

        <div className="page-header">
        <div className="content-container">
        <h1 className="page-header__title">Viewing <span>{viewingText}</span> totalling <span>{totalFormated}</span></h1>
        <div className="page-header__actions">
        <Link className="button" to="/create">Add Expense</Link>
        </div>
        </div>
        </div>
    )
}

const connectPropToStore = (state) => ({
    expenses: selectExpenses(state.expenses, state.filter),
    
})

export default connect(connectPropToStore)(ExpensesSummary)