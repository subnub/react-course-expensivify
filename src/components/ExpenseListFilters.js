import React from "react";
import {connect} from "react-redux";
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from "../actions/filters"
import {DateRangePicker} from "react-dates"

class ExpenesListFilters extends React.Component {

    state = {
        calanderFocus: null
    };

    onDatesChange = ({startDate, endDate}) => {

        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }

    onFocusChange = (calanderFocus) => {

        this.setState(() => ({calanderFocus}))
    }

    render() {
        return (
        <div className="content-container">
            <div className="input-group">

                <div className="input-group__item">
                    <input className="text-input" 
                    placeholder="Search Expenses"
                    type="text" value={this.props.filter.text} onChange={(e) => {
                        this.props.dispatch(setTextFilter(e.target.value))
                    }}/>
                </div>

                <div className="input-group__item">
                    <select className="select" 
                    value={this.props.filter.sortBy} onChange={(e) => {
                
                        e.target.value === "date" ? this.props.dispatch(sortByDate()) 
                        : this.props.dispatch(sortByAmount());
            
                    }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                    </select>
                </div>

                <div className="input-group__item">
                    <DateRangePicker 
                    startDate={this.props.filter.startDate}
                    endDate={this.props.filter.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calanderFocus}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                    />
                </div>
            
            </div>

        </div> 
        )
    }
}


const mapStateToProps = (state) => {

    return {
        filter: state.filter
    }
}

export default connect(mapStateToProps)(ExpenesListFilters);