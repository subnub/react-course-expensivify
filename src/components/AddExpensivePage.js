import React from "react";
import ExpenseForm from "./ExpenseForm"
import {connect} from "react-redux"
import {startAddExpense} from "../actions/expenses"

export class AddExpensivePage extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense)
        this.props.history.push("/dashboard")
    }

    render() {
        return (
        <div>

            <div className="page-header">
                <div className="content-container">
                    <h1>Add Expense</h1>
                </div>
            </div>

            <div className="content-container">
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => {dispatch(startAddExpense(expense))}
})

export default connect(undefined, mapDispatchToProps)(AddExpensivePage);