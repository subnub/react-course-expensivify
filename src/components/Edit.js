import React from "react";
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm"
import {editExpense, removeExpense, startEditExpense, startRemoveExpense} from "../actions/expenses"

export class Edit extends React.Component {

    constructor(props) {
        super(props)
    }

    onSubmit = (expense) => {

        console.log("submit", this.props.expense.id)
        this.props.dispatch(startEditExpense(this.props.expense.id,expense))
        this.props.history.push("/dashboard")
    }

    onClick = (e) => {
        this.props.dispatch(startRemoveExpense({id: this.props.expense.id}))
        this.props.history.push("/dashboard")
    }

    render() {

        return (
            <div>

                <div className="page-header">
                <div className="content-container">
                    <h1>Edit Expense</h1>
                </div>
                </div>

                <div className="content-container">

                    <ExpenseForm expense={this.props.expense} onSubmit={(expense) => {
                        this.onSubmit(expense)
                    }}/>
                    
                    <div>
                        <button className="button button--secondary" onClick={(e) => {
                            this.onClick(e)
                        }}>Remove</button>
                    </div>
    
                </div>
                
            </div>
        )
    }
}

// const Edit = (props) => {
    
//     console.log(props);

    

//     return (
//     <div>
//         <ExpenseForm expense={props.expense} onSubmit={(expense) => {
//             console.log("submit", props.expense.id)
//             props.dispatch(editExpense(props.expense.id,expense))
//             props.history.push("/")
//         }}/>
//         <button onClick={(e) => {
//             props.dispatch(removeExpense({id: props.expense.id}))
//             props.history.push("/")
//         }}>Remove</button>
//     </div>
// )}

const mapToStore = (state, props) => {

    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}

const mapDispatchToProps = (dispatch) => ({
    
})

export default connect(mapToStore)(Edit);