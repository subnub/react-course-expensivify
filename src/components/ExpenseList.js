import React from "react";
import {connect} from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenes from "../selectors/expense";

export const ExpenseList = (props) => (

    <div>

        {props.expenses.length === 0 ?

            (<p>No Expensess</p>) 
            
            : 

            (props.expenses.map((expense) => {
                console.log(expense)
                return (
                    <ExpenseListItem key={expense.id} {...expense}/>
                )
            }))}

        
    </div>
)

const mapStataeToProps = (state) => {

    console.log(state);
    return {
        expenses: selectExpenes(state.expenses, state.filter)
    }
}

export default connect(mapStataeToProps)(ExpenseList);

