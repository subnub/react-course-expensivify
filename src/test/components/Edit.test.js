import {Edit} from "../../components/Edit";
import {shallow} from "enzyme";
import React from "react";
import expenses from "../fixtures/expenses";

let editExpense, removeExpense, history, onSubmit,wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = jest.fn();
    onSubmit = jest.fn();

    wrapper = shallow(<Edit 
        editExpense={editExpense}
        removeExpense={removeExpense}
        history={history}
        onSubmit={onSubmit}
        expense={expenses[1]}/>)
})

test("Should render edit expense with an item", () => {

    expect(wrapper).toMatchSnapshot();        

})

// test("Should call onSubmit with proper data", () => {
//     // const onSubmitSpy = jest.fn();
//     const expense = expenses[1]
//     // const wrapper = shallow(<Edit expense={expense}/>)

//     wrapper.find("ExpenseForm").prop("onSubmit")(expense)

//     expect(onSubmitSpy).toHaveBeenLastCalledWith({
//         desc: expense.desc,
//         note: expense.note,
//         amount: expense.amount,
//         createdAt: expense.createdAt
//     })
// })