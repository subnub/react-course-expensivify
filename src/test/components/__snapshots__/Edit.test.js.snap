// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expense with an item 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 190500,
        "createdAt": -345600000,
        "desc": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
